lib/membrane_element_aac/encoder.ex

```diff
@@ -0,0 +1,248 @@
+    end
+  end
+
+  def handle_caps(:input, caps, _ctx, %{input_caps: stored_caps}) do
+    raise """
+    Received caps #{inspect(caps)} are different then defined in options #{inspect(stored_caps)}.
+    If you want to allow converter to accept different input caps dynamically, use `nil` as input_caps.
+    """
+  end
+
+  @impl true
+  def handle_process(:input, %Buffer{payload: data}, _ctx, state) do
+    %{native: native, queue: queue} = state
+
+    to_encode = queue <> data
+
+    with {:ok, {encoded_buffers, bytes_used}} when bytes_used > 0 <-
+           encode_buffer(to_encode, native) do
+      <<_handled::binary-size(bytes_used), rest::binary>> = to_encode
+
+      buffer_actions = [buffer: {:output, encoded_buffers}]
+
+      {{:ok, buffer_actions ++ [redemand: :output]}, %{state | queue: rest}}
+    else
+      {:ok, {[], 0}} -> {:ok, %{state | queue: to_encode}}
+      {:error, reason} -> {{:error, reason}, state}
+    end
+  end
+
+  @impl true
+  def handle_event(:input, %EndOfStream{}, _ctx, state) do
+    %{native: native, queue: queue} = state
+
+    if queue != <<>>,
+      do: warn("Processing queue is not empty, but EndOfStream event was received")
+
+    actions = [event: {:output, %EndOfStream{}}, notify: {:end_of_stream, :input}]
+
+    with {:ok, encoded_frame} <- Native.encode_frame(<<>>, native) do
+      buffer_actions = [buffer: {:output, %Buffer{payload: encoded_frame}}]
+
+      {{:ok, buffer_actions ++ actions}, state}
+    else
+      {:error, :no_data} ->
+        {{:ok, actions}, state}
+
+      {:error, reason} ->
+        {{:error, reason}, state}
+    end
+  end
+
+  def handle_event(pad, event, ctx, state) do
+    super(pad, event, ctx, state)
+  end
+
+  # Initialize buffer encoding
+  defp encode_buffer(buffer, native) do
+    raw_frame_size = @aac_frame_size * @default_channels * @sample_size
+
+    encode_buffer(buffer, native, [], 0, raw_frame_size)
+  end
+
+  # Encode a single frame if buffer contains at least one frame
+  defp encode_buffer(buffer, native, acc, bytes_used, raw_frame_size)
+       when byte_size(buffer) >= raw_frame_size do
+    <<raw_frame::binary-size(raw_frame_size), rest::binary>> = buffer
+
+    with {:ok, encoded_frame} <- Native.encode_frame(raw_frame, native) do
+      encoded_buffer = %Buffer{payload: encoded_frame}
+
+      # Continue encoding the rest until no more frames are available in the queue
+      encode_buffer(
+        rest,
+        native,
+        [encoded_buffer | acc],
+        bytes_used + raw_frame_size,
+        raw_frame_size
+      )
+    else
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  # Not enough samples for a frame
+  defp encode_buffer(_native, _partial_buffer, acc, bytes_used, _raw_frame_size) do
+    # Return accumulated encoded frames
+    {:ok, {acc |> Enum.reverse(), bytes_used}}
+  end
+
+  defp mk_native(channels, sample_rate, aot) do
+    with {:ok, channels} <- validate_channels(channels),
```

Comment:
1. For the caps received on `:input` pad the validation against `@supported_caps` is done automatically
2. You can validate the caps received via options in `handle_init` (and it would be done only once)

Reply:
@bblaszkow06 
Right, but I don't think we can do the validation in `handle_init` as the `input_caps` could be specified as `nil` (both the whole `input_caps` object as well as its internal properties) and received in `handle_caps`, which means that this would either have to handle `nil`s as separate edge case or allow `nil` in the validation function.
If you think we should do that anyway, I'm fine with changing that though.